packages/backend-function/src/function_env_translator.ts

```diff
@@ -1,79 +1,111 @@
     private readonly functionEnvironmentProp: Required<FunctionProps>['environment'],
     private readonly backendSecretResolver: BackendSecretResolver
   ) {
-    const secretPlaceholderText = '<value will be resolved during runtime>';
-    const amplifySecretPaths = 'AMPLIFY_SECRET_PATHS';
-    const secretPathEnvVars: AmplifySecretPaths = {};
-
     for (const [key, value] of Object.entries(this.functionEnvironmentProp)) {
-      if (key === amplifySecretPaths) {
+      if (key === this.amplifySsmEnvConfigKey) {
         throw new Error(
-          `${amplifySecretPaths} is a reserved environment variable name`
+          `${this.amplifySsmEnvConfigKey} is a reserved environment variable name`
         );
       }
       if (typeof value !== 'string') {
         const { branchSecretPath, sharedSecretPath } =
           this.backendSecretResolver.resolvePath(value);
-        this.environmentRecord[key] = secretPlaceholderText;
-        secretPathEnvVars[branchSecretPath] = {
+        this.lambda.addEnvironment(key, this.ssmValuePlaceholderText);
+        this.ssmEnvVars[branchSecretPath] = {
           name: key,
           sharedPath: sharedSecretPath,
         };
-        this.secretPaths.push(branchSecretPath, sharedSecretPath);
+        this.ssmPaths.push(branchSecretPath, sharedSecretPath);
       } else {
-        this.environmentRecord[key] = value;
+        this.lambda.addEnvironment(key, value);
       }
     }
 
-    this.environmentRecord[amplifySecretPaths] =
-      JSON.stringify(secretPathEnvVars);
-  }
-
-  getEnvironmentRecord = () => {
-    return this.environmentRecord;
-  };
-
-  getSecretPolicyStatement = (): iam.PolicyStatement | undefined => {
-    if (this.secretPaths.length === 0) {
-      return;
-    }
-
-    const stack = Stack.of(this.scope);
-
-    const resourceArns = this.secretPaths.map(
-      (path) => `arn:aws:ssm:${stack.region}:${stack.account}:parameter${path}`
+    // add an environment variable for ssm parameter metadata that is resolved after initialization but before synth is finalized
+    this.lambda.addEnvironment(
+      this.amplifySsmEnvConfigKey,
+      Lazy.string({
+        produce: () => JSON.stringify(this.ssmEnvVars),
+      })
     );
 
-    return new iam.PolicyStatement({
-      effect: iam.Effect.ALLOW,
-      actions: ['ssm:GetParameters'],
-      resources: resourceArns,
+    this.lambda.node.addValidation({
+      validate: () => {
+        // only add the ssm access policy if there are ssm paths
+        if (this.ssmPaths.length > 0) {
+          const ssmAccessPolicy = new PolicyStatement({
+            effect: Effect.ALLOW,
+            actions: ['ssm:GetParameters'],
+            resources: this.ssmPaths
+              .map((path) => (path.startsWith('/') ? path.slice(1) : path)) // the Arn formatter will add a leading slash between the resource and resourceName
+              .map((path) =>
```

Comment:
why don't we combine the two loops?

Reply:
It seemed more readable to separate out sanitizing the path from arn generation